Add local thread pool to CCheckQueue

hebasto opened the pull request:
This PR:
- gets rid of `boost::thread_group` in the `CCheckQueue` class
- allows thread safety annotation usage in the `CCheckQueue` class
- is alternative to #14464 (https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-616618525, https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-617291612)

Also, with this PR (I hope) it could be easier to resurrect a bunch of brilliant ideas from #9938.

Related: #17307

MarcoFalke commented:
Concept ACK on getting rid of boost in consensus code. This pull looks similar to #14464?

MarcoFalke commented:
> Also, with this PR (I hope) it could be easier to resurrect a bunch of brilliant ideas from #9938.

cc @JeremyRubin

hebasto commented:
@MarcoFalke 
> Concept ACK on getting rid of boost in consensus code. This pull looks similar to #14464?

Thanks for pointing to #14464. I'll look into it.

hebasto commented:
Updated 9764d81e1f000d1bd3c03ac6f6f4361c4aa84ee9  -> f4cd37e93a522d5ab2bc872d129a0142be773a7d ([pr18710.01](https://github.com/hebasto/bitcoin/commits/pr18710.01) -> [pr18710.02](https://github.com/hebasto/bitcoin/commits/pr18710.02), [diff](https://github.com/hebasto/bitcoin/compare/pr18710.01..pr18710.02)):

- fixed linter issue with boost headers

JeremyRubin commented:
Nice! From a simple code review PoV it looks good.

I remember there being a bunch of subtle nasties in the specifics of using std::threads v.s. boost threads around interrupt system and the details of how condition variables work. I can't remember what they all were though now. I think #14464 closed because of that...


But generally huge concept ack on anything that makes #9938 more likely to move forward! Note though that the cuckoocache sort of stole it's thunder -- a lot of the contention in the CheckQueue was around the sigcache locking and the cuckoocache made that stuff way faster so it's not obvious #9938 will show as impressive gains, and the trade offs for consensus code may be less worth it. Can discuss more via IRC; I'd love to get some of the improvements through in any case.

hebasto commented:
@MarcoFalke 
> Concept ACK on getting rid of boost in consensus code. This pull looks similar to #14464?

The main difference from #14464 is additional atomic `m_destruction_in_progress` that replaces the boost "interrupt" feature: https://github.com/bitcoin/bitcoin/blob/f4cd37e93a522d5ab2bc872d129a0142be773a7d/src/checkqueue.h#L96-L97 and https://github.com/bitcoin/bitcoin/blob/f4cd37e93a522d5ab2bc872d129a0142be773a7d/src/checkqueue.h#L172-L173

hebasto commented:
Updated f4cd37e93a522d5ab2bc872d129a0142be773a7d  -> b02bc2583e421dbca35a5652c650ae861230f480 ([pr18710.02](https://github.com/hebasto/bitcoin/commits/pr18710.02) -> [pr18710.03](https://github.com/hebasto/bitcoin/commits/pr18710.03), [diff](https://github.com/hebasto/bitcoin/compare/pr18710.02..pr18710.03)):

- picked some ideas from #14464
- reduced scope of the PR (some commits moved into a separate PR)
- OP updated (now every commit passes unit tests)

hebasto commented:
Benchmark results on my machine:
- master (a998c5185bc7fc2c7e22312fac60175cb2869bdd)
```
# Benchmark, evals, iterations, total, min, max, median
CCheckQueueSpeedPrevectorJob, 100, 1400, 96.3747, 0.000677139, 0.000695471, 0.000688339
```

- this PR (b02bc2583e421dbca35a5652c650ae861230f480)

```
# Benchmark, evals, iterations, total, min, max, median
CCheckQueueSpeedPrevectorJob, 100, 1400, 77.4393, 0.000545271, 0.00057892, 0.000552496
```

hebasto commented:
Rebased b02bc2583e421dbca35a5652c650ae861230f480 -> https://github.com/bitcoin/bitcoin/commit/f16f7f79dcd5d6c1b55966317b8762abeb5561c3 ([pr18710.03](https://github.com/hebasto/bitcoin/commits/pr18710.03) -> [pr18710.04](https://github.com/hebasto/bitcoin/commits/pr18710.04)) due to the conflict with #18575.

MarcoFalke commented on src/checkqueue.h:
```diff
@@ -132,16 +139,31 @@ class CCheckQueue
+        assert(m_worker_threads.empty());
+        m_stopped = false;
+        m_worker_threads.reserve(threads_num);
+        for (auto n = 0; n < threads_num; ++n) {
```
```suggestion
        for (int n = 0; n < threads_num; ++n) {
```

is shorter and more precise

hebasto replied:
Agree.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-619008172).

MarcoFalke commented on src/checkqueue.h:
```diff
@@ -132,16 +139,31 @@ class CCheckQueue
+        if (threads_num <= 0) return;
+        assert(m_worker_threads.empty());
+        m_stopped = false;
+        m_worker_threads.reserve(threads_num);
```
is this useful in any way?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-619008172).

MarcoFalke commented on src/checkqueue.h:
```diff
@@ -132,16 +139,31 @@ class CCheckQueue
+    //! Create a pool of new worker threads.
+    void StartWorkerThreads(const int threads_num)
+    {
+        if (threads_num <= 0) return;
```
Is this needed? the for loop will never run anyway when this is true.

hebasto replied:
The idea was to prevent `m_stopped = false;`

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-619008172).

MarcoFalke commented on src/checkqueue.h:
```diff
@@ -95,6 +101,7 @@ class CCheckQueue
                         return fRet;
                     }
                     nIdle++;
+                    if (m_stopped) return true; // TODO: This return value is unused.
```
Also, shouldn't the `if (m_stopped)` be *after* the condition variable `wait`? Otherwise notifying it will first empty the queue.
Finally, `m_stopped` could be renamed to `m_request_stop` and only briefly set in `StopWorkerThreads` to true and then immediately after to false. No other member functions (like StartWorkerThreads) touch it

hebasto replied:
> Otherwise notifying it will first empty the queue.

Correct. This is done intentionally as I think that worker threads should exit with a clean check queue, no?

hebasto replied:
> Also, shouldn't the `if (m_stopped)` be _after_ the condition variable `wait`?

This wouldn't work for the master thread that could never return.

hebasto replied:
> Finally, `m_stopped` could be renamed to `m_request_stop` and only briefly set in `StopWorkerThreads` to true and then immediately after to false.

How this approach could work when `StopWorkerThreads()` is called during check queue being processed? I think `m_stopped` should keep its state in that case.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-619008172).

MarcoFalke commented on src/test/checkqueue_tests.cpp:
```diff
@@ -9,7 +9,7 @@
 
 #include <boost/test/unit_test.hpp>
-#include <boost/thread.hpp>
+#include <boost/thread/thread.hpp>
```
1c7fdb71b32bd91f6ae10313169d9eec63d46e3e: This commit should probably go last, because some of the changes are removed earlier anyway

hebasto replied:
In that case in the commit _"bench: Use CCheckQueue local thread pool"_ the `lint-includes.sh` linter will produce:
```
Good job! The Boost dependency "boost/thread/thread.hpp" is no longer used.
Please remove it from EXPECTED_BOOST_INCLUDES in ./test/lint/lint-includes.sh
to make sure this dependency is not accidentally reintroduced.
```
which seems a bit falsely.

MarcoFalke replied:
I hope that no one is running the linters on every commit. They are meant to warn authors and reviewers as well as maintainers of code smell in a pull request as a whole. They are not meant as unit tests that absolutely must pass on every commit.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-619008172).

MarcoFalke commented on src/test/checkqueue_tests.cpp:
```diff
@@ -373,71 +347,7 @@ BOOST_AUTO_TEST_CASE(test_CheckQueue_FrozenCleanup)
-/** Test that CCheckQueueControl is threadsafe */
-BOOST_AUTO_TEST_CASE(test_CheckQueueControl_Locks)
-{
-    auto queue = MakeUnique<Standard_Queue>(QUEUE_BATCH_SIZE);
```
d64dd5ffee064b0b943b7e00b16b0ad28ac65474: Why remove this test?

hebasto replied:
I have no idea how to pass a customized callback into the `CCheckQueue::StartWorkerThreads(int)`.

JeremyRubin replied:
Usually you want to use a callable template parameter...
```c++
template<typename Callable>
function(int x, Callable&& function) {
    function(x);
}
function(1, [](int x) { return x});
function(1, [](int x) { });
```

hebasto replied:
@JeremyRubin Thank you for your suggestion :)

Sorry for my broken English, but I meant do we really need a unit test if the `CCheckQueue` class has no interface to test?

JeremyRubin replied:
Sorry to just be responding on this:

I think if there isn't an interface for something, but you can expose one to make testing more robust, it's worthwhile.

E.g., the CheckQueue does not need to be templated as we know a concrete single type for it, but templating it allows us to mock out the job type and probe the internal execution, which is a good thing.

What's the missing interface here? I don't see one being used in the code for this test.

hebasto replied:
@MarcoFalke @JeremyRubin 
I was confused about this test (I don't know the exact reason of that confusion now).
Test has been [restored](https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-639992778).

MarcoFalke commented on src/checkqueue.h:
```diff
@@ -33,34 +30,29 @@ template <typename T>
-    //! Worker threads block on this when out of work
-    boost::condition_variable condWorker;
-
-    //! Master thread blocks on this when out of work
```
629dae24827121412fb24d31ac19a918ade1db3f: why remove the documentation?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-619008172).

maflcko reviewed (approved):
Approach ACK f16f7f79dcd5d6c1b55966317b8762abeb5561c3 🌘

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Approach ACK f16f7f79dcd5d6c1b55966317b8762abeb5561c3 🌘
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhhpwv+NPSS8v5sdXJqBUVegg4/8o1tsvoNpU9cE80q4qYqkmrP2V55QWbkG2VI
NE6+E0XSgTx8UuESI+9YEJ9+8gH4jCgNIqDMS4dqG+uPeeTXJUKn39Jn59YuaJGQ
yWCCO4BZXqW/DWCLVpcDINXGtlZgFR3sv53RiFPWw+NNU5aJYTFssv9KJ8nBQVko
SXNKSNzFxqx2oafuIaF+Ikrp+HbPzwE7Xrk2byS9Od1sA8d/LGkbhxBB4j32L4NT
8NNUBKdata4zjzkQCXHIrBYNxJxy8/qK45lMO4WjiApxyA5kbbOZ20v9eyh8oF5I
4hsvQN1Bpvdm8hK3oqxcLzZ2/QO6SjkNnVslNlFeYubasFMoTGS7n0SMqgmqgIbW
qi9sMpsuMEBov6wq5kCIpWTBwHIuDMeRbjo/dOkJcTNqJHu8+s83aqD8P3MbTa0y
ykeCbWwB8k0v5SPGhKMRx5sUMOGHQBWboDRzDJmNF909W+Yd3DL+C0qONk6vvMSy
03UYQCgR
=16Sm
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `fd6cc8b78bfb432d1fffbc1706dd13c7f3563a0d2caef6be9b45688ccd1b15ad  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108fd6cc8b78bfb432d1fffbc1706dd13c7f3563a0d2caef6be9b45688ccd1b15adf0102d304663044c55e8f77b4602ebfa03ed08fff0109ebd3db5bd9aa9a63414bcce3367652008f120cc004781951cb0a81f4af96994b49a60d7b31644cb7e0771c90f5ced06df15bd08f1045ea22acff00824b4f8e8d750ab290083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010cf3c1caf96986585a8b6aa554098f83f08f1045ea22ad0f008dd2cf8baddcc1cd50083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010ebfe40ecb50d112a4d80a978df473ec708f020832dce27a967e201eed36948203e515f41ebe3e832a425aa831200eb5353637608f0205ebbe934329fbb44ffb3de970cace10ff805230c4fa64b853ad243645c47847a08f1045ea22ad0f008ba51217e883891080083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f6e4377f5c867e9810fc716f45dfa3a008f1206455ed265297873929f4db46643c8208cbaf2c8c2e5f81996b72cc7cba81e20d08f1045ea22ad0f00851c172a228f0ffa10083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

hebasto commented:
Updated f16f7f79dcd5d6c1b55966317b8762abeb5561c3 -> 7a7e346a1d153fc7367bca5a1926881af65279ec ([pr18710.04](https://github.com/hebasto/bitcoin/commits/pr18710.04) -> [pr18710.05](https://github.com/hebasto/bitcoin/commits/pr18710.05), [diff](https://github.com/hebasto/bitcoin/compare/pr18710.04..pr18710.05)):

- addressed all of @MarcoFalke's comments

MarcoFalke commented:
In the last commit: Could replace the last occurrence of `thread.hpp` with `#include <boost/thread/shared_mutex.hpp>` and kill boost thread from the linter?

hebasto commented:
Updated 7a7e346a1d153fc7367bca5a1926881af65279ec -> 61739174c67a773d124734ab6ec3385657fe3ef8 ([pr18710.05](https://github.com/hebasto/bitcoin/commits/pr18710.05) -> [pr18710.06](https://github.com/hebasto/bitcoin/commits/pr18710.06), [diff](https://github.com/hebasto/bitcoin/compare/pr18710.05..pr18710.06)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-619023235):

> In the last commit: Could replace the last occurrence of `thread.hpp` with `#include <boost/thread/shared_mutex.hpp>` and kill boost thread from the linter?

hebasto commented:
@MarcoFalke It seems there is a thread issue on [windows](https://travis-ci.org/github/bitcoin/bitcoin/jobs/679063276)...

hebasto commented:
Updated 61739174c67a773d124734ab6ec3385657fe3ef8 -> b3ac0c3d15a729cdce5f15b918947037905b59dc ([pr18710.06](https://github.com/hebasto/bitcoin/commits/pr18710.06) -> [pr18710.07](https://github.com/hebasto/bitcoin/commits/pr18710.07), [diff](https://github.com/hebasto/bitcoin/compare/pr18710.06..pr18710.07)):

- [FIXUP](https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-619126820):
> It seems there is a thread issue on windows: https://travis-ci.org/github/bitcoin/bitcoin/jobs/679063276 and https://bitcoinbuilds.org/?job=90df9b0f-d918-45a1-b27d-0ae6d4cec303 ...

hebasto commented:
Reworked dac216cee9e2e8fece089f67cf84f80f6890541f:

- fixed data race conditions
- dropped "refactor: Change return type of CCheckQueue::Loop()" commit

promag reviewed (commented):
Code review ACK dac216cee9e2e8fece089f67cf84f80f6890541f.

Is f9c3bf45c13c147cb8d66d156519af021d66218e really a refactor due to changes between boost and std threads?

hebasto commented:
> Is [f9c3bf4](https://github.com/bitcoin/bitcoin/commit/f9c3bf45c13c147cb8d66d156519af021d66218e) really a refactor due to changes between boost and std threads?

I think it is, as the code in the pre- f9c3bf45c13c147cb8d66d156519af021d66218e state does not rely on `boost::thread` features, no?

hebasto commented:
Updated dac216cee9e2e8fece089f67cf84f80f6890541f -> 0b5e2a5528fae627e7be98ac231bfb9ae24cd504 ([pr18710.08](https://github.com/hebasto/bitcoin/commits/pr18710.08) -> [pr18710.09](https://github.com/hebasto/bitcoin/commits/pr18710.09), [diff](https://github.com/hebasto/bitcoin/compare/pr18710.08..pr18710.09)):

- added missed `GUARDED_BY` thread-safety annotation

hebasto commented:
Rebased 0b5e2a5528fae627e7be98ac231bfb9ae24cd504 -> 82d5e650ee2b72d1447cbfc7bf1bb107dfd390b6 ([pr18710.09](https://github.com/hebasto/bitcoin/commits/pr18710.09) -> [pr18710.10](https://github.com/hebasto/bitcoin/commits/pr18710.10)) on top of the #18881.

hebasto commented:
Rebased 82d5e650ee2b72d1447cbfc7bf1bb107dfd390b6 -> 46838e94da7c54389bce1fb7c0951ab338dcc304 ([pr18710.10](https://github.com/hebasto/bitcoin/commits/pr18710.10) -> [pr18710.11](https://github.com/hebasto/bitcoin/commits/pr18710.11)) on top of the #16127.

hebasto commented:
Rebased 46838e94da7c54389bce1fb7c0951ab338dcc304 -> 8457827f784e0e2586b19e78785d143b8eeb9513 ([pr18710.11](https://github.com/hebasto/bitcoin/commits/pr18710.11) -> [pr18710.12](https://github.com/hebasto/bitcoin/commits/pr18710.12)) on top of the #18792.

hebasto commented:
Rebased 8457827f784e0e2586b19e78785d143b8eeb9513 -> b392ef178725795d9e38a4dd000b20ee47c0d60a ([pr18710.12](https://github.com/hebasto/bitcoin/commits/pr18710.12) -> [pr18710.13](https://github.com/hebasto/bitcoin/commits/pr18710.13)) on top of the #19142.

hebasto commented:
Rebased b392ef178725795d9e38a4dd000b20ee47c0d60a -> eee80a1c36d886583e387934abce6dac9b3117b5 ([pr18710.13](https://github.com/hebasto/bitcoin/commits/pr18710.13) -> [pr18710.14](https://github.com/hebasto/bitcoin/commits/pr18710.14)) due to the conflict with #18758.

JeremyRubin commented:
utack eee80a1

Failure is a build pause, job needs restart. I'm still not sure why the test needed to be removed, maybe you can explain that more clearly.

hebasto commented:
Updated eee80a1c36d886583e387934abce6dac9b3117b5 -> ec4401480246f60c6b22d98d205a509a181010d7 ([pr18710.14](https://github.com/hebasto/bitcoin/commits/pr18710.14) -> [pr18710.15](https://github.com/hebasto/bitcoin/commits/pr18710.15), [diff](https://github.com/hebasto/bitcoin/compare/pr18710.14..pr18710.15)):

- restored the `test_CheckQueueControl_Locks` unit test that was removed by mistake

hebasto commented:
@JeremyRubin 
> Failure is a build pause, job needs restart.

No chance, unfortunately. See: #19171.
The #18731 has been built on top of this PR successfully.

JeremyRubin commented:
utACK ec44014


Annoying note:

As noted in https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-616979595, Would be good if someone with access to some of the different platforms/arches we support can do benching/real network profiling to just to confirm there's no perf regression on the condition variable and mutex behavior. Otherwise seems ready to merge.

see https://codesequoia.wordpress.com/2013/03/27/condition-variables-performance-of-boost-win32-and-the-c11-standard-library/ for some context, some std synchronization primitives are worse on some platforms and can have more variance.

But to be clear, these potential regressions would probably be small compared to other improvements realizable in the checkqueue design, so I don't think it's a blocker for accepting these changes but we do want to know if merging this means we should be prioritizing some optimizations to neutralize any degradation.

laanwj commented:
Concept ACK, will review

adamjonas commented:
To summarize the state of review for this PR:

Concept ACK (laanwj), Approach ACK (MarcoFalke), Code review ACK (Promag), utACK (JeremyRubin) with no NACKs or blocking concerns raised.

@JeremyRubin [asked](https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-640095382) if someone "with access to some of the different platforms/arches we support can do benching/real network profiling to just to confirm there's no perf regression on the condition variable and mutex behavior."

[Benchmarks from the author](https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-617446268):
master (a998c51)
```
# Benchmark, evals, iterations, total, min, max, median
CCheckQueueSpeedPrevectorJob, 100, 1400, 96.3747, 0.000677139, 0.000695471, 0.000688339
```
this PR (b02bc25 - pre-rebase)
```
# Benchmark, evals, iterations, total, min, max, median
CCheckQueueSpeedPrevectorJob, 100, 1400, 77.4393, 0.000545271, 0.00057892, 0.000552496
```

hebasto commented:
Rebased ec4401480246f60c6b22d98d205a509a181010d7 -> e7893d2e522477374ef9bf01997e9776c86b9813 ([pr18710.15](https://github.com/hebasto/bitcoin/commits/pr18710.15) -> [pr18710.16](https://github.com/hebasto/bitcoin/commits/pr18710.16)) due to the conflicts with #18637 and #19589.

hebasto commented:
Updated benchmarks (Linux Mint 20, x86_64):
- master (2f71a1ea35667b3873197201531e7ae198ec5bf4)
```
# Benchmark, evals, iterations, total, min, max, median
CCheckQueueSpeedPrevectorJob, 5, 1400, 4.31038, 0.000612868, 0.000620909, 0.000614335
```

- this PR (e7893d2e522477374ef9bf01997e9776c86b9813)
```
# Benchmark, evals, iterations, total, min, max, median
CCheckQueueSpeedPrevectorJob, 5, 1400, 3.67213, 0.000517622, 0.000531811, 0.000523883
```

hebasto commented:
Rebased e7893d2e522477374ef9bf01997e9776c86b9813 -> a57a0e526c94b35b1b018fde5a8a2ea46fca8899 ([pr18710.16](https://github.com/hebasto/bitcoin/commits/pr18710.16) -> [pr18710.17](https://github.com/hebasto/bitcoin/commits/pr18710.17)) due to the conflicts with #18011 and #19604.

Updated benchmarks (Linux Mint 20, x86_64):
- master (a63a26f042134fa80356860c109edb25ac567552)
```
|              ns/job |               job/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              189.96 |        5,264,153.75 |    4.1% |      0.01 | `CCheckQueueSpeedPrevectorJob`
```

- this PR (a57a0e526c94b35b1b018fde5a8a2ea46fca8899)
```
|              ns/job |               job/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              157.27 |        6,358,598.51 |    1.9% |      0.01 | `CCheckQueueSpeedPrevectorJob
```

martinus commented:
Hi! I only have access to two different computers, an Intel i7-8700, and an Intel Celeron N3050. Compiled with g++ 10.1.0.

# i7-8700 (12 threads)
master f7c73b03d975a72f609ded2bbe250c1c8a76a944, i7-8700:

|              ns/job |               job/s |    err% |         ins/job |         cyc/job |    IPC |        bra/job |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|              582.25 |        1,717,469.22 |    0.5% |          267.01 |          347.19 |  0.769 |          50.65 |    3.7% |      0.02 | `CCheckQueueSpeedPrevectorJob`

PR a57a0e526c94b35b1b018fde5a8a2ea46fca8899, i7-8700:

|              ns/job |               job/s |    err% |         ins/job |         cyc/job |    IPC |        bra/job |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|              371.47 |        2,692,026.21 |    0.6% |          260.57 |          334.12 |  0.780 |          49.78 |    3.7% |      0.01 | `CCheckQueueSpeedPrevectorJob`

PR is much faster here, quite an improvement here! Results are also relatively stable, at least stable enough to see that the PR is faster on average.

# Intel Celeron N3050 (2 threads)

Here the runtime fluctuates so much that it doesn't make much sense to show the raw numbers. Especially the PR seems much more prone to fluctuations. To get a better understanding of the fluctuations, I've set `epoch(100)`, and generated boxplots with

```cpp
std::ofstream html("pr.html");
bench.render(ankerl::nanobench::templates::htmlBoxplot(), html);
```

Then I get two html's for each run, which I've merged by hand:

![newplot(2)](https://user-images.githubusercontent.com/14386/89100570-343d3c80-d3f8-11ea-929c-acd42f71b0a8.png)

The PR seems to be slower here, but this might also be an artifact of the benchmark. I don't think this benchmark is very good.

martinus commented:
I ran the benchmark again on the Intel Celeron N3050, but with `BATCHES=5001` instead of 101. The benchmark takes a long time, and is a bit more stable. It still fluctuates by a factor of 4. On average the PR is slower than master on this machine.

![newplot(4)](https://user-images.githubusercontent.com/14386/89105933-b7c15280-d425-11ea-89f0-791dd2e57fd8.png)

hebasto commented on src/checkqueue.h:
```diff
@@ -62,8 +64,11 @@ class CCheckQueue
     //! The maximum number of elements to be processed in one batch
     const unsigned int nBatchSize;
 
+    std::vector<std::thread> m_worker_threads;
```
Why? `m_worker_threads` is accessed from one thread only.

hebasto commented:
Updated a57a0e526c94b35b1b018fde5a8a2ea46fca8899 -> e084c3b6051fc596d0b9a38ae7a5da58e3637a31 ([pr18710.17](https://github.com/hebasto/bitcoin/commits/pr18710.17) -> [pr18710.18](https://github.com/hebasto/bitcoin/commits/pr18710.18), [diff](https://github.com/hebasto/bitcoin/compare/pr18710.17..pr18710.18)):

- addressed @promag's nits

- added commit b8e267134431db97e853cddd32fdc41716d3414e "bench: Prevent thread oversubscription"

Benchmarks on my machine (Linux Mint 20, x86_64):
- master (a63a26f042134fa80356860c109edb25ac567552)
```
|              ns/job |               job/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              191.58 |        5,219,878.03 |    2.1% |      0.01 | `CCheckQueueSpeedPrevectorJob`
```

- master + b8e267134431db97e853cddd32fdc41716d3414e (without thread oversubscription)
```
|              ns/job |               job/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              182.46 |        5,480,666.18 |    2.0% |      0.01 | `CCheckQueueSpeedPrevectorJob`
```

- this PR (e084c3b6051fc596d0b9a38ae7a5da58e3637a31)
```
|              ns/job |               job/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              150.84 |        6,629,325.36 |    1.1% |      0.01 | `CCheckQueueSpeedPrevectorJob`
```

hebasto commented on src/checkqueue.h:
```diff
@@ -132,16 +141,34 @@ class CCheckQueue
+        }
+        assert(m_worker_threads.empty());
+        for (int n = 0; n < threads_num; ++n) {
+            m_worker_threads.emplace_back([this, n]() {
```
[Updated](https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-667653036).

hebasto commented on src/checkqueue.h:
```diff
@@ -132,16 +141,34 @@ class CCheckQueue
+            nTotal = 0;
+            fAllOk = true;
+        }
+        assert(m_worker_threads.empty());
```
[Updated](https://github.com/bitcoin/bitcoin/pull/18710#issuecomment-667653036).

hebasto commented:
@martinus

There are some concerns about thread oversubscription impact on benchmarking.

Mind repeating your benchmarks on 2-thread system with updated branch?

martinus commented:
Sure, I ran it again. i7-8700 is faster too. I also tested it with `queue.StartWorkerThreads(1);`, which is by far the fastest for that benchmark. It seems that the work units are far too small to be able to reasonably distribute the work. So I am not sure if this benchmark is much useful.

|              ns/job |               job/s |    err% |         ins/job |         cyc/job |    IPC |        bra/job |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|              582.25 |        1,717,469.22 |    0.5% |          267.01 |          347.19 |  0.769 |          50.65 |    3.7% |      0.02 | master f7c73b03d975a72f609ded2bbe250c1c8a76a944
|              371.47 |        2,692,026.21 |    0.6% |          260.57 |          334.12 |  0.780 |          49.78 |    3.7% |      0.01 | PR a57a0e526c94b35b1b018fde5a8a2ea46fca8899
|              351.57 |        2,844,406.84 |    1.7% |          261.09 |          332.71 |  0.785 |          49.90 |    3.6% |      0.11 | PR e084c3b6051fc596d0b9a38ae7a5da58e3637a31
|              131.95 |        7,578,398.22 |    1.1% |          262.19 |          252.28 |  1.039 |          50.15 |    3.8% |      0.04 | queue.StartWorkerThreads(1);

Updated graph with `BATCHES = 5001` on the Celeron N3050:

![newplot(5)](https://user-images.githubusercontent.com/14386/89121056-e4c04480-d4bb-11ea-9415-d921e03da576.png)

hebasto commented:
@martinus Great! Thanks very much!

Mind including benchmarks for b8e267134431db97e853cddd32fdc41716d3414e as well, as this commit changes the benchmark itself?

martinus commented:
If I understand it correcly, the benchmark queues `PrevectorJob` as as the job to simulate. But the jobs itself do nothing, they simply return true:

```cpp
bool operator()()  {
    return true;
}
```
I'd say the job should do some reasonable amount of work so using multiple threads actually make any sense.

Anyways, here is an update with b8e2671 included:

i7-8700:

|              ns/job |               job/s |    err% |         ins/job |         cyc/job |    IPC |        bra/job |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|              582.25 |        1,717,469.22 |    0.5% |          267.01 |          347.19 |  0.769 |          50.65 |    3.7% |      0.02 | master f7c73b03d975a72f609ded2bbe250c1c8a76a944
|              371.47 |        2,692,026.21 |    0.6% |          260.57 |          334.12 |  0.780 |          49.78 |    3.7% |      0.01 | PR a57a0e526c94b35b1b018fde5a8a2ea46fca8899
|              351.57 |        2,844,406.84 |    1.7% |          261.09 |          332.71 |  0.785 |          49.90 |    3.6% |      0.11 | PR e084c3b6051fc596d0b9a38ae7a5da58e3637a31
|              131.95 |        7,578,398.22 |    1.1% |          262.19 |          252.28 |  1.039 |          50.15 |    3.8% |      0.04 | queue.StartWorkerThreads(1);
|              548.98 |        1,821,565.90 |    2.9% |          267.48 |          334.32 |  0.800 |          50.76 |    3.6% |      0.02 | b8e267134


Celeron N3050:

![newplot(7)](https://user-images.githubusercontent.com/14386/89121718-a3329800-d4c1-11ea-8bd1-66da20619696.png)


They behave very similar, e084c3b6051fc596d0b9a38ae7a5da58e3637a31 is a bit faster than b8e267134.

hebasto commented:
This PR consist of two distinct changes now:
- master -> b8e267134431db97e853cddd32fdc41716d3414e
  - prevented thread oversubscription during benchmarking
  - benchmark change only
  - benchmarking results variance becomes lower
- b8e267134431db97e853cddd32fdc41716d3414e -> e084c3b6051fc596d0b9a38ae7a5da58e3637a31
  - dropped `boost::thread_group`
  - benchmarking results mean improved on all tested platforms (including 2-threaded Celeron N3050)

What did I miss?

hebasto commented:
To make reviewing easier, the first commit is split out to #19710.

hebasto committed:
refactor: Use member initializers in CCheckQueue

hebasto committed:
Add local thread pool to CCheckQueue

hebasto committed:
test: Use CCheckQueue local thread pool

hebasto committed:
bench: Use CCheckQueue local thread pool

hebasto committed:
refactor: Drop boost::thread stuff in CCheckQueue

hebasto commented:
Rebased e084c3b6051fc596d0b9a38ae7a5da58e3637a31 -> fa815a305108bc8b25ac468abaec7fa60c261f41 ([pr18710.18](https://github.com/hebasto/bitcoin/commits/pr18710.18) -> [pr18710.19](https://github.com/hebasto/bitcoin/commits/pr18710.19)) on top of #19710 to prevent a merge conflict.

hebasto commented:
Rebased fa815a305108bc8b25ac468abaec7fa60c261f41 -> 89c4f7591e02df4fd36071995740113c1c06db3d ([pr18710.19](https://github.com/hebasto/bitcoin/commits/pr18710.19) -> [pr18710.20](https://github.com/hebasto/bitcoin/commits/pr18710.20)) due to the conflict with #19710.

hebasto commented:
Rebased 89c4f7591e02df4fd36071995740113c1c06db3d -> bb6fcc75d1ec94b733d1477c816351c50be5faf9 ([pr18710.20](https://github.com/hebasto/bitcoin/commits/pr18710.20) -> [pr18710.21](https://github.com/hebasto/bitcoin/commits/pr18710.21)) due to the conflict with #19927.

jamesob commented:
Concept ACK, will plan to review.

LarryRuane reviewed (commented):
nit: remove the `#include <boost/thread/thread.hpp>` from checkqueue_tests.cpp.

Looks good. I ran the affected tests (and single-stepped through most of the new test code), and verified using debugger that a mainnet node correctly reaches the real [script checking code](https://github.com/bitcoin/bitcoin/blob/dff0f6f753eafd932d7d65fbfa33585f620e8e54/src/validation.cpp#L1402). 
ACK bb6fcc75d1ec94b733d1477c816351c50be5faf9

hebasto commented:
@LarryRuane 
> nit: remove the `#include <boost/thread/thread.hpp>` from checkqueue_tests.cpp.

`boost::thread_group` is still used in the `test_CheckQueueControl_Locks`.

LarryRuane commented:
> `boost::thread_group` is still used in the `test_CheckQueueControl_Locks`

It builds for me without that include. Oh, I see, it's included indirectly via `test/util/setup_common.h`, thanks.

laanwj commented:
Thanks for getting rid of the one-to-last use of `boost::thread_group`.

Code review ACK bb6fcc75d1ec94b733d1477c816351c50be5faf9

jonatack reviewed (commented):
Code review ACK bb6fcc75d1ec94b733d1477c816351c50be5faf9 and verified rebase to master builds cleanly with unit/functional tests green
